insolar/bus/bus.go

```diff
@@ -0,0 +1,163 @@
+// limitations under the License.
+//
+
+package bus
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/ThreeDotsLabs/watermill/message/router/middleware"
+	"github.com/insolar/insolar/instrumentation/inslogger"
+	"github.com/pkg/errors"
+)
+
+const (
+	// TopicOutgoing is topic for external calls
+	TopicOutgoing = "TopicOutgoing"
+
+	// TopicIncoming is topic for incoming calls
+	TopicIncoming = "TopicIncoming"
+
+	// MetaPulse is key for Pulse
+	MetaPulse = "pulse"
+
+	// MetaType is key for Type
+	MetaType = "type"
+
+	// MetaReceiver is key for Receiver
+	MetaReceiver = "receiver"
+
+	// MetaSender is key for Sender
+	MetaSender = "sender"
+)
+
+//go:generate minimock -i github.com/insolar/insolar/insolar/bus.Sender -o ./ -s _mock.go
+
+// Sender interface sends messages by watermill.
+type Sender interface {
+	// Send an `Message` and get a `Reply` or error from remote host.
+	Send(ctx context.Context, msg *message.Message) (<-chan *message.Message, func())
+}
+
+type lockedReply struct {
+	mutex    sync.RWMutex
+	messages chan *message.Message
+
+	once sync.Once
+	done chan struct{}
+}
+
+// Bus is component that sends messages and gives access to replies for them.
+type Bus struct {
+	pub     message.Publisher
+	timeout time.Duration
+
+	repliesMutex sync.RWMutex
+	replies      map[string]*lockedReply
+}
+
+// NewBus creates Bus instance with provided values.
+func NewBus(pub message.Publisher) *Bus {
+	return &Bus{
+		timeout: time.Minute * 10,
+		pub:     pub,
+		replies: make(map[string]*lockedReply),
+	}
+}
+
+func (b *Bus) removeReplyChannel(ctx context.Context, id string, reply *lockedReply) {
+	reply.once.Do(func() {
+		close(reply.done)
+
+		b.repliesMutex.Lock()
+		defer b.repliesMutex.Unlock()
+		delete(b.replies, id)
+
+		inslogger.FromContext(ctx).Infof("close reply channel for message with correlationID %s", id)
+
+		reply.mutex.Lock()
+		close(reply.messages)
+		reply.mutex.Unlock()
+	})
+}
+
+// Send a watermill's Message and return channel for replies.
+func (b *Bus) Send(ctx context.Context, msg *message.Message) (<-chan *message.Message, func()) {
+	id := watermill.NewUUID()
+	middleware.SetCorrelationID(id, msg)
+	reply := &lockedReply{
+		messages: make(chan *message.Message),
+		done:     make(chan struct{}),
+	}
+	b.repliesMutex.Lock()
+	defer b.repliesMutex.Unlock()
+
+	err := b.pub.Publish(TopicOutgoing, msg)
+	if err != nil {
+		inslogger.FromContext(ctx).Errorf("can't publish message to %s topic: %s", TopicOutgoing, err.Error())
+		return nil, nil
+	}
+
+	b.replies[id] = reply
+
+	c := func(ctx context.Context, b *Bus, reply *lockedReply) func() {
+		return func() {
+			b.removeReplyChannel(ctx, id, reply)
+		}
+	}(ctx, b, reply)
+
+	go func(c func()) {
+		select {
+		case <-reply.done:
+			inslogger.FromContext(msg.Context()).Infof("reply channel for message with correlationID %s was closed", id)
```

Comment:
This log msg sounds the same as in removeReplyChannel(). Make them more specific